async_fuse/src/byte_slice.rs

```diff
@@ -0,0 +1,208 @@
+use anyhow::Context;
+use std::ffi::{CStr, OsStr};
+use std::mem;
+use std::slice;
+
+use super::protocol::FuseAbiData;
+
+/// # Safety
+/// + ensure that `bytes.as_ptr()` is well-aligned for T.
+/// + ensure that `bytes.len()` is equal to `size_of::<T>()`.
+unsafe fn transmute_ref_unchecked<T: FuseAbiData>(bytes: &[u8]) -> &T {
+    &*(bytes.as_ptr().cast::<T>())
+}
+
+/// # Safety
+/// + ensure that `bytes.as_ptr()` is well-aligned for T.
+/// + ensure that `bytes.len()` is a multiple of `size_of::<T>()`.
+unsafe fn transmute_slice_unchecked<T: FuseAbiData>(bytes: &[u8]) -> &[T] {
+    let len = bytes.len() / mem::size_of::<T>();
+    slice::from_raw_parts(bytes.as_ptr().cast::<T>(), len)
+}
+
+/// A slice of bytes which is used for parsing FUSE abi data.
+#[derive(Debug)]
+pub(crate) struct ByteSlice<'a> {
+    data: &'a [u8],
+}
+
+impl<'a> ByteSlice<'a> {
+    pub(crate) fn new(data: &'a [u8]) -> ByteSlice<'a> {
+        ByteSlice { data }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub(crate) fn fetch_all(&mut self) -> &'a [u8] {
+        mem::replace(&mut self.data, &[])
+    }
+
+    pub(crate) fn fetch_bytes(&mut self, amt: usize) -> anyhow::Result<&'a [u8]> {
+        if amt > self.data.len() {
+            anyhow::bail!(
+                "no enough bytes to fetch, remaining {} bytes but to fetch {} bytes",
+                self.data.len(),
+                amt
+            );
+        }
+        let (bytes, remain) = self.data.split_at(amt);
+        self.data = remain;
+        Ok(bytes)
+    }
+
+    pub(crate) fn fetch<T: FuseAbiData>(&mut self) -> anyhow::Result<&'a T> {
+        let elem_len: usize = mem::size_of::<T>();
+
+        if (self.data.as_ptr() as usize) % mem::align_of::<T>() != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to type {}, pointer={:p} is not a multiple of alignment={}",
+                std::any::type_name::<T>(),
+                self.data.as_ptr(),
+                mem::align_of::<T>()
+            );
+        }
+
+        let bytes: &[u8] = self.fetch_bytes(elem_len).with_context(|| {
+            format!(
+                "failed to build FUSE request payload type {}",
+                std::any::type_name::<T>(),
+            )
+        })?;
+
+        let ret: &T = unsafe { transmute_ref_unchecked(bytes) };
+        Ok(ret)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn fetch_all_as_slice<T: FuseAbiData>(&mut self) -> anyhow::Result<&'a [T]> {
+        let elem_len: usize = mem::size_of::<T>();
+
+        if self.data.len() % elem_len != 0 {
+            anyhow::bail!(
+                "failed to convert bytes to a slice of type={}, \
+                 the total bytes length={} % the type size={} is nonzero",
```

Comment:
more indentation (4 spaces) for the second line of the string:
```
"failed to convert bytes to a slice of type={}, \
    the total bytes length={} % the type size={} is nonzero",
```

Reply:
done